From: Dirk Buchwalder <buchwalder@posteo.de>
Date: Sun, 08 Jan 2023 19:34:16 +0200
Subject: [PATCH] ath11k: rm Register shutdown handler for WCN6750
as this leads to a q6v5 crash during reboot on ipq60xx

https://lore.kernel.org/linux-wireless/20221009221453.1216158-23-sashal@kernel.org/

--- a/drivers/net/wireless/ath/ath11k/ahb.c
+++ b/drivers/net/wireless/ath/ath11k/ahb.c
@@ -1211,10 +1211,20 @@ err_core_free:
 	return ret;
 }
 
-static void ath11k_ahb_remove_prepare(struct ath11k_base *ab)
+static int ath11k_ahb_remove(struct platform_device *pdev)
 {
+	struct ath11k_base *ab = platform_get_drvdata(pdev);
 	unsigned long left;
 
+	if (test_bit(ATH11K_FLAG_QMI_FAIL, &ab->dev_flags)) {
+		ath11k_ahb_power_down(ab);
+		ath11k_debugfs_soc_destroy(ab);
+		ath11k_qmi_deinit_service(ab);
+		goto qmi_fail;
+	}
+
+	reinit_completion(&ab->driver_recovery);
+
 	if (test_bit(ATH11K_FLAG_RECOVERY, &ab->dev_flags)) {
 		left = wait_for_completion_timeout(&ab->driver_recovery,
 						   ATH11K_AHB_RECOVERY_TIMEOUT);
@@ -1224,13 +1234,9 @@ static void ath11k_ahb_remove_prepare(st
 
 	set_bit(ATH11K_FLAG_UNREGISTERING, &ab->dev_flags);
 	cancel_work_sync(&ab->restart_work);
-	cancel_work_sync(&ab->qmi.event_work);
-}
-
-static void ath11k_ahb_free_resources(struct ath11k_base *ab)
-{
-	struct platform_device *pdev = ab->pdev;
 
+	ath11k_core_deinit(ab);
+qmi_fail:
 	ath11k_ahb_free_irq(ab);
 	ath11k_hal_srng_deinit(ab);
 	ath11k_ahb_release_smp2p_handle(ab);
@@ -1238,47 +1244,10 @@ static void ath11k_ahb_free_resources(st
 	ath11k_ce_free_pipes(ab);
 	ath11k_core_free(ab);
 	platform_set_drvdata(pdev, NULL);
-}
-
-static int ath11k_ahb_remove(struct platform_device *pdev)
-{
-	struct ath11k_base *ab = platform_get_drvdata(pdev);
-
-	if (test_bit(ATH11K_FLAG_QMI_FAIL, &ab->dev_flags)) {
-		ath11k_ahb_power_down(ab);
-		ath11k_debugfs_soc_destroy(ab);
-		ath11k_qmi_deinit_service(ab);
-		goto qmi_fail;
-	}
-
-	ath11k_ahb_remove_prepare(ab);
-	ath11k_core_deinit(ab);
-
-qmi_fail:
-	ath11k_ahb_free_resources(ab);
 
 	return 0;
 }
 
-static void ath11k_ahb_shutdown(struct platform_device *pdev)
-{
-	struct ath11k_base *ab = platform_get_drvdata(pdev);
-
-	/* platform shutdown() & remove() are mutually exclusive.
-	 * remove() is invoked during rmmod & shutdown() during
-	 * system reboot/shutdown.
-	 */
-	ath11k_ahb_remove_prepare(ab);
-
-	if (!(test_bit(ATH11K_FLAG_REGISTERED, &ab->dev_flags)))
-		goto free_resources;
-
-	ath11k_core_deinit(ab);
-
-free_resources:
-	ath11k_ahb_free_resources(ab);
-}
-
 static struct platform_driver ath11k_ahb_driver = {
 	.driver         = {
 		.name   = "ath11k",
@@ -1286,7 +1255,6 @@ static struct platform_driver ath11k_ahb
 	},
 	.probe  = ath11k_ahb_probe,
 	.remove = ath11k_ahb_remove,
-	.shutdown = ath11k_ahb_shutdown,
 };
 
 static int ath11k_ahb_init(void)
--- a/drivers/net/wireless/ath/ath11k/core.c
+++ b/drivers/net/wireless/ath/ath11k/core.c
@@ -1692,7 +1692,6 @@ void ath11k_core_pre_reconfigure_recover
 	wake_up(&ab->wmi_ab.tx_credits_wq);
 	wake_up(&ab->peer_mapping_wq);
 
-	reinit_completion(&ab->driver_recovery);
 }
 
 static void ath11k_core_post_reconfigure_recovery(struct ath11k_base *ab)
